roachprod: load balancer from pgurl command

herkolategan opened the pull request:
Previously, the load balancer IP or URL can only be retrieved using the expanders. This change adds the ability to also get the URL from the `roachprod` pgurl command by using the `lb` node selector similar to how it is done with expansion.

Epic: None
Release Note: None

herkolategan committed:
roachprod: load balancer from pgurl command

Previously, the load balancer IP or URL can only be retrieved using the
expanders. This change adds the ability to also get the URL from the `roachprod`
pgurl command by using the `lb` node selector similar to how it is done with
expansion.

Epic: None
Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123206)
<!-- Reviewable:end -->

srosenberg commented on pkg/roachprod/install/nodes.go:
```diff
@@ -44,7 +44,9 @@ func ListNodes(s string, numNodesInCluster int) (Nodes, error) {
-	if s == "all" {
+	// "L" is a special value that also returns all nodes, but implies a load
+	// balancer should be used when applicable.
+	if s == "all" || s == "L" {
```
Why not `LB`? `L` doesn't seem like a memorable mnemonic unlike `all`. Should it also be case-insensitive?

herkolategan replied:
I matched this to how the expanders also select the load balancer vs. node IPs. I can update both. I'll maybe add it to the command help as well to make it clear as an option.

herkolategan committed:
roachprod: use 'lb' to denote load balancer

Previously, `L` was used in the `roachprod` expanders and other place to select
load balancers. This change updates it to `LB` or `lb` to be more explicit. The
help for commands have also been updated to inform the user that an option
exists to select a load balancer.

Epic: None
Release Note: None

srosenberg reviewed (approved):
Thanks for switching to the `lb` mnemonic; I promise it will be easier to remember (for me) :) The rest looks good, but do update the commit msg. to replace `L` with `lb`.

renatolabs commented on pkg/roachprod/install/nodes.go:
```diff
@@ -44,7 +44,9 @@ func ListNodes(s string, numNodesInCluster int) (Nodes, error) {
 
-	if s == "all" {
+	// "lb" is a special value that also returns all nodes, but implies a load
+	// balancer should be used when applicable.
```
Why "when applicable"? I don't see logic below to handle the case where a load balancer is not available (would we want that?)

herkolategan replied:
True, I'll update the comment to be more concise.

renatolabs commented on pkg/roachprod/roachprod.go:
```diff
@@ -184,6 +184,13 @@ func newCluster(
 	return c, nil
 }
 
+// preferLoadBalancer determines if the node selector section in the cluster
```
I guess this is a similar point, but `prefer` suggests that actually returning a LB IP is optional, but it seems we return an error below if we can't find a LB.

I actually prefer this semantics (error returned if LB requested but doesn't exist) unless there's strong reason to provide a fallback. In that case, maybe `requestedLoadBalancer` might be a better name.

renatolabs commented on pkg/roachprod/install/expander.go:
```diff
@@ -22,8 +22,8 @@ import (
 var parameterRe = regexp.MustCompile(`{[^{}]*}`)
-var pgURLRe = regexp.MustCompile(`{pgurl(:[-,0-9]+|:L)?(:[a-z0-9\-]+)?(:[0-9]+)?}`)
-var pgHostRe = regexp.MustCompile(`{pghost(:[-,0-9]+|:L)?(:[a-z0-9\-]+)?(:[0-9]+)?}`)
+var pgURLRe = regexp.MustCompile(`{pgurl(:[-,0-9]+|:(?i)lb)?(:[a-z0-9\-]+)?(:[0-9]+)?}`)
```
At some point it would be nice to document these with examples, these regexes are getting complicated 😄

herkolategan replied:
True, these are becoming a bit over-encumbered.

herkolategan commented:
TFTRs!

bors r=srosenberg,renatolabs
